### PR TITLE
#122 Don't call setters on omitted fields during PATCH

### LINKED ISSUE #122: save() doesn't adhere to JSON API PATCH semantics
## Context

In Katharsis `save()` is used for creation and updating.

When used for updating during a PATCH request Katharsis sets **all** the properties on the JsonApiResource POJO in contradiction to the JSON API standard.

> If a request does not include all of the attributes for a resource, the server MUST interpret the missing attributes as if they were included with their current values. The server MUST NOT interpret missing attributes as null values.

## Reproduce

Given a `Person` JsonApiResource with the business rules:
1. Creating a `Person` with no initial attributes is permitted
2. Changing a `Person's attributes to null is **not** permitted.

Creating a person with no initial attributes is also possible because of the default constructor required by frameworks (hibernate, katharsis, etc).

```java
@JsonApiResource(type = "persons")
public class Person()
{
    private Long id;

    private Integer age;

    private String name;

    // default constructor
    public Person() {}

    public Person(String name, Integer age) {
        this.name = name;
        this.age = age;
    }

    @JsonApiId
    public Long getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        if (name == null) throw new IllArgEx("...");
        this.name = name;
    }

    public Integer getAge() {
        return age;
    }

    public void setAge(Integer age) {
        if (age == null) throw new IllArgEx("...");
        this.age = age;
    }
}
```

Now, I have an instance of Person: `id: 5, name: Jon, age: null`. His age has never been set, so it is null on the server, but I would like to change his name to `John` so I PATCH with:

```json
{
  "data": {
    "type":"persons",
    "id":5
    "attributes": {
      "name":"John"
    }
  }
}
```

**Expected Result**: 

The resource's `setName("John")` method is called, and `setAge()` is never called.

**Actual Result**:
Katharsis calls `setName("John")`  properly, but also calls `setAge(null)` during the PATCH request and triggers the violation of business rule #2.

## Solution?

I'm not sure how to solve this problem but, the reason is that object `DataBody` contains the null field age and `ResourceAttributeBridge.setProperties(...)` will try to set all the attributes it finds in the `DataBody` object.

One could use this nasty workaround:
```java
public void setAge(Integer age) {
    if (this.age != null && age == null) throw new IllArgExc("...");
    this.age = age;
}
```

It's not sensible to overwrite all fields of an object with the values it already holds, because:
* A setter may also do complicated operations like parsing the input, transforming it, etc
* It leads to the type of workarounds above which needlessly pollute the codebsae with noise unrelated to business logic.

### PR DESCRIPTION
**Pull request by Ramblurr**
*(trying out the new PR template from #243)*

## Description

During a PATCH request with a partial attributes payload, the setters for the omitted attributes would still be called.

1. This is redundant, as the existing values are already loaded via the `findOne` call
2. It breaks developer expectations that setters of a JsonApiResource are only called for meaningful/semantic business reasons, i.e., that a request was submitted that results in a field being modified.

## Motivation and Context

Fixes #122. See discussion and examples there.

## How Has This Been Tested?

I've added a special test method in `ResourcePatchTest` that explicitly tests this behavior.

1. `Task` now has an additional attribute called `category`.
2. The category setter throws an exception if null is passed to it (the idea being that null category is not allowed for some business rule reason)

For testing:
* A Task object is created that has a null category and a name set.
* A PATCH request is made updating the name, but omitting the category.

Before my patch, this test would fail as the exception in the setCategory method would be thrown.
After my patch, the exception is not triggered and setCategory is never invoked, which is the expected behavior

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the code style of this project: no doc changes
- [x] My change requires a change to the documentation: no
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: katharsis-core/src/main/java/io/katharsis/dispatcher/controller/resource/ResourcePatch.java [hunk 1]
```diff
@@ -77,12 +78,22 @@ public BaseResponseContext handle(JsonPath jsonPath, QueryAdapter queryAdapter,
...
-            // get the JSON form the request and deserialize into a map
+            // deserialize the request JSON's attributes object into a map
             String attributesAsJson = objectMapper.writeValueAsString(dataBody.getAttributes());
             Map<String,Object> attributesFromRequest = objectMapper.readValue(attributesAsJson, Map.class);;
+
+            // remove attributes that were omitted in the request
+            Iterator<String> it = attributesToUpdate.keySet().iterator();
+            while(it.hasNext()) {
+            	String key = it.next();
+			    if(attributesFromRequest.containsKey(key))
```
Comment (chb0github) [inline katharsis-core/src/main/java/io/katharsis/dispatcher/controller/resource/ResourcePatch.java]: If you flip this predicate around you can simply get rid of the `continue` statement

File: katharsis-core/src/test/java/io/katharsis/resource/mock/models/Task.java [hunk 2]
```diff
@@ -77,6 +73,15 @@ public void setName(@SuppressWarnings("SameParameterValue") String name) {
         this.name = name;
     }
 
+    public String getCategory() {
+        return category;
+    }
+
+    public void setCategory(String category) {
+    	if(category == null) throw new IllegalArgumentException("Category cannot be set to null!");
```
Comment (chb0github) [inline katharsis-core/src/test/java/io/katharsis/resource/mock/models/Task.java]: As these are only example classes, this is fine. But really, a class just for testing should be created.

Comment (Ramblurr) [inline katharsis-core/src/test/java/io/katharsis/resource/mock/models/Task.java]: I agree with you in general. At first I created a new Resource + Repository for this test, however I wanted to ensure that no other existing tests broke. The only way to do that without writing/copying new tests was to modify an already heavily tested Resource. Hence the use of `Task`.

**Comment by Ramblurr** [comment]
Updated the loop, and added the test case I had forgotten to include in the commit.

**Comment by Ramblurr** [comment]
@chb0github  @remmeier  any other comments before merging?

**Comment by chb0github** [comment]
I am good

**Pull request merged**